pkg/controller/miganalytic/restic_executor.go

```diff
@@ -0,0 +1,150 @@
+}
+
+// DF given a podRef and a list of volumes, runs df command, returns with structured command context
+// any errors running the df command are suppressed here. DFCommand.stdErr field should be used to determine failure
+func (r *ResticDFCommandExecutor) DF(podRef *corev1.Pod, persistentVolumes []MigAnalyticPersistentVolumeDetails) DFCommand {
+	// TODO: use the appropriate block size based on PVCs
+	dfCmd := DFCommand{
+		BaseLocation: "/host_pods",
+		BlockSize:    DecimalSIMega,
+		StdOut:       "",
+		StdErr:       "",
+	}
+	cmdString := dfCmd.PrepareDFCommand(persistentVolumes)
+	restCfg := r.Client.RestConfig()
+	podCommand := pods.PodCommand{
+		Pod:     podRef,
+		RestCfg: restCfg,
+		Args:    cmdString,
+	}
+	err := podCommand.Run()
+	if err != nil {
+		log.Info(
+			fmt.Sprintf("df command inside pod %s returned non-zero error code", podRef.Name))
+	}
+	dfCmd.StdErr = podCommand.Err.String()
+	dfCmd.StdOut = podCommand.Out.String()
+	return dfCmd
+}
+
+// getResticPodForNode lookup Restic Pod ref in local cache
+func (r *ResticDFCommandExecutor) getResticPodForNode(nodeName string) *corev1.Pod {
+	if podRef, exists := r.ResticPodReferences[nodeName]; exists {
+		return podRef
+	}
+	return nil
+}
+
+// loadResticPodReferences load Restic Pod refs in-memory
+func (r *ResticDFCommandExecutor) loadResticPodReferences() error {
+	if r.ResticPodReferences == nil {
+		r.ResticPodReferences = make(map[string]*corev1.Pod)
+	}
+	resticPodList := corev1.PodList{}
+	labelSelector := client.InNamespace(r.Namespace).MatchingLabels(
+		map[string]string{
+			ResticPodLabelKey: ResticPodLabelValue})
+	err := r.Client.List(context.TODO(), labelSelector, &resticPodList)
+	if err != nil {
+		return err
+	}
+	for i := range resticPodList.Items {
+		if resticPodList.Items[i].Spec.NodeName != "" {
+			r.ResticPodReferences[resticPodList.Items[i].Spec.NodeName] = &resticPodList.Items[i]
+		}
+	}
+	return nil
+}
+
+// Execute given a map node->[]pvc, runs Df command for each, returns list of structured df output per pvc
+func (r *ResticDFCommandExecutor) Execute(pvcNodeMap map[string][]MigAnalyticPersistentVolumeDetails) ([]DFOutput, error) {
+	gatheredData := []DFOutput{}
+	err := r.loadResticPodReferences()
+	if err != nil {
+		return gatheredData, err
+	}
+	// dfOutputs for n nodes
+	dfOutputs := make(map[string]DFCommand, len(pvcNodeMap))
+	var waitGroup sync.WaitGroup
+	var mutex sync.Mutex
+	waitGroup.Add(len(pvcNodeMap))
+	// run df concurrently for 'n' nodes
+	for node := range pvcNodeMap {
+		resticPodRef := r.getResticPodForNode(node)
+		// if no Restic pod is found for this node, all PVCs on this node are skipped
+		if resticPodRef == nil {
+			for _, pvc := range pvcNodeMap[node] {
+				dfOutput := DFOutput{
+					IsError:   true,
+					Name:      pvc.Name,
+					Namespace: pvc.Namespace,
+				}
+				gatheredData = append(gatheredData, dfOutput)
+			}
+			continue
+		}
+		go func(n string, podRef *corev1.Pod, outputs map[string]DFCommand) {
```

Comment:
it does not appear that there is any limit or buffer, so lets say we have 500 nodes we will have 500 threads, and then we will quickly make 500 calls and then they will block until the map is updated. 

Not only will this not really gain us as much value from the async (I think right?) it will spam the API server will a ton of requests really quickly. Assuming that each call takes some period of time, we may want to consider making a buffered channel executor that also parses the output from below in each thread and the return is pvcDFInfo?

WDYT? too much over engineering?

Reply:
@shawn-hurley good point. I think it's fair to assume a scenario for 500 nodes. I can use your help implementing some sort of a buffered channel. Haven't really done that.

Reply:
+1 good idea